src: signature: Add new feature 'kw signature'

Lincoln-Yuji opened the pull request:
### Purpose
Often a kernel developer has to perform some operations over commit or patch trailer lines. Adding tags such as 'Reviewed-by' or 'Acked-by' is a common task and sometimes it might be required to do this over multiple commits or patches. Also another common one is the 'Fixes' tag, that must follow a specific format.

This feature helps on these cases. It uses 'git rebase', 'git commit --amend' and 'git interpret-trailers' as its backend in order to make these tasks quicker to complete.

### Inspiration
Both #1049 and #1051 represent similar issues, regarding trailer line manipulation of commits and patches. The idea is to deliver a new feature that can make such tasks easier and quicker to deal with.

### Some known limitations
There is no way to "revert" any change made by this command through kw. For example, if the commit log changes, there is no command to undo the previous operation. User will have to manually use `git reset --hard` if changes are not what they expected.

This feature currently does not provide ways to configure conditional actions, which is possible via `git config` as shown by Melissa here: #1049

rodrigosiqueira commented on src/handle_trailer.sh:
```diff
@@ -0,0 +1,169 @@
+        options_values['TRAILER_TAG']='Fixes:'
+        options_values['TRAILER_VALUE']="$(str_strip "$2")"
+
+        # Check if given value is a valid commit reference
+        if [[ $(git cat-file -t "${options_values['TRAILER_VALUE']}" 2> /dev/null) != 'commit' ]]; then
+          options_values['ERROR']='Invalid commit reference with --add-fixes option: '
+          options_values['ERROR']+="${options_values['TRAILER_VALUE']}"
+          return 22 # EINVAL
+        fi
+        # The 'Fixes:' trailer line must follow a format defined by
+        # Linux Kernel developers. Example:
+        # Fixes: e21d2170f366 ("video: remove unnecessary platform_set_drvdata()")
+        options_values['TRAILER_VALUE']=$(git log -1 "${options_values['TRAILER_VALUE']}" \
+          --oneline --format="%h (\\\"%s\\\")" \
+          --abbrev-commit --abbrev=12)
```
It would be better if this manipulation were moved to a single function and called from the main functions instead of here.

rodrigosiqueira commented on src/handle_trailer.sh:
```diff
@@ -0,0 +1,169 @@
+function handle_trailer_help()
+{
+  if [[ "$1" == --help ]]; then
+    include "$KW_LIB_DIR/help.sh"
```
String concatenation.

rodrigosiqueira commented on tests/unit/handle_trailer_test.sh:
```diff
@@ -0,0 +1,260 @@
+Signed-off-by: kw <kw@kw>"
+
+CORRECT_REVIEWED_LOG="Signed-off-by: kw <kw@kw>
+Reviewed-by: John Doe <johndoe@community.com>
```
I think `community.com` is a valid email; just use some random name.

rodrigosiqueira commented on tests/unit/handle_trailer_test.sh:
```diff
@@ -0,0 +1,260 @@
+# Correct trailers for --add-reviewed-by and -r
+CORRECT_REVIEWED_COMMIT="Signed-off-by: kw <kw@kw>
+Reviewed-by: John Doe <johndoe@community.com>
+
```
Why do we need these extra spaces (for all the options), and why is the signed-off-by duplicate? This does not look correct.

Lincoln-Yuji replied:
I believe the confusion here happened because of our poor choice of names for the variables and lack of better comments to explain what these variables are holding. Those variables are holding the trailer lines of multiple commits. For instance, when using the command `git local --max-count 2 --format="%(trailers)"` the output follows this format:

```
Signed-off-by: Name <mail>                              <- Most recent commit
Reviewed-by: Another Name <another mail>
                                                        <- Empty line splitting trailers from different commits
Signed-off-by: Name <mail>                              <- Second most recent commit trailers
Reviewed-by: Some Name <some mail>
```

This is why the `Signed-off` appears duplicated in multiple outputs and there are these extra spaces in all the options. We hope this clarified the issue. We've added some comments and changed the name of some of these variables to make it easier to understand.

Also please tell us if we misunderstood the problem here. Perhaps we could be missing something important

Lincoln-Yuji replied:
I believe the confusion here happened because of our poor choice of names for the variables and lack of better comments to explain what these variables are holding. Those variables are holding the trailer lines of multiple commits. For instance, when using the command `git log --max-count 2 --format="%(trailers)"` the output follows this format:

```
Signed-off-by: Name <mail>                  <- Most recent commit
Reviewed-by: Another Name <another mail>
                                            <- Empty line splitting trailers from different commits
Signed-off-by: Name <mail>                  <- Second most recent commit trailers
Reviewed-by: Some Name <some mail>
```

This is why `Signed-off` appears duplicated in multiple outputs and there are these extra spaces in all the options. We hope this clarified the issue. We've added some comments and changed the name of some of these variables to make it easier to understand.

Also please tell us if we misunderstood the problem here. Perhaps we could be missing something important

rodrigosiqueira commented on tests/unit/handle_trailer_test.sh:
```diff
@@ -0,0 +1,260 @@
+{
+  local -r current_path="$PWD"
+
+  cp -r tests/unit/samples/handle_trailer_patch_samples/*.patch "$SHUNIT_TMPDIR"/
```
Use full parameter names and add quotes around the path.

rodrigosiqueira commented on tests/unit/handle_trailer_test.sh:
```diff
@@ -0,0 +1,260 @@
+  mkdir fs
+  touch fs/some_file
+  git add fs/
+  git commit --quiet -s -m "fs: some_file: Fill file" -m "First"
```
https://kworkflow.org/content/codingstyle.html#avoid-short-options

rodrigosiqueira commented on tests/unit/handle_trailer_test.sh:
```diff
@@ -0,0 +1,260 @@
+  git commit --quiet -s -m "fs: some_file: Fill file" -m "First"
+
+  # Simulate wrong change
+  echo 'Wrong text' > fs/some_file
```
https://kworkflow.org/content/codingstyle.html#use-printf-instead-of-echo

rodrigosiqueira commented on tests/unit/handle_trailer_test.sh:
```diff
@@ -0,0 +1,260 @@
+
+function tearDown()
+{
+  rm -rf "$SHUNIT_TMPDIR"
+  mkdir -p "$SHUNIT_TMPDIR"
```
https://kworkflow.org/content/codingstyle.html#avoid-short-options

rodrigosiqueira commented on tests/unit/handle_trailer_test.sh:
```diff
@@ -0,0 +1,260 @@
+function test_handle_trailer_main_patch()
+{
+  local original_dir="$PWD"
+  cd "$SHUNIT_TMPDIR" || {
+    fail "($LINENO) It was not possible to move to temporary directory"
+    return
+  }
```
Why not move this cd to the setup and the cd at the end of this test to the teardown?

Also: https://kworkflow.org/content/codingstyle.html#string-concatenation

rodrigosiqueira commented on tests/unit/handle_trailer_test.sh:
```diff
@@ -0,0 +1,260 @@
+    return
+  }
+
+  # Testing --add-reviewed-by and -r with patch
+  handle_trailer_main --add-reviewed-by "John Doe <johndoe@community.com>" patch_model.patch
+  assertFileEquals 'patch_model.patch' 'patch_model_reviewed.patch'
+  git restore patch_model.patch
+
+  handle_trailer_main -r "John Doe <johndoe@community.com>" patch_model.patch
+  assertFileEquals 'patch_model.patch' 'patch_model_reviewed.patch'
+  git restore patch_model.patch
+
+  # Testing --add-acked-by and -a with patch
+  handle_trailer_main --add-acked-by "Michael Doe <michaeldoe@community.com>" patch_model.patch
+  assertFileEquals 'patch_model.patch' 'patch_model_acked.patch'
+  git restore patch_model.patch
```
Maybe split this test into small tests where each part test a specif argument, for example, tests for acked-by and others for reviewed-by

rodrigosiqueira commented on tests/unit/samples/handle_trailer_patch_samples/patch_model.patch:
```diff
@@ -0,0 +1,33 @@
+ullam corporis suscipit laboriosam, nisi ut aliquid ex ea
+commodi consequatur.
+
+Signed-off-by: kw <kw@example.net>
```
example.net sounds like a real email; try to use something more random.

rodrigosiqueira reviewed (changes requested):
Hi,
Sorry for the delay tbh; I was thinking a lot about this feature... It is a very good feature, but I'm still not sure if the name is good enough or if we can put it under other feature... Anyway, since I don't have any good options for this, I decided to go on and review it.

Again, thanks a lot for your PR.

Lincoln-Yuji commented on documentation/man/features/kw-handle-trailer.rst:
```diff
@@ -0,0 +1,86 @@
+OPTIONS
+=======
+-r, \--add-reviewed-by:
+  Adds a **Reviewed-by** trailer line to either commits or patch files.
```
If I understood correctly, we need to change this line (and every other with the same problem I assume) explaining the option more precisely:
"Adds a <tag> trailer line to either patch files, patchsets or series"

Is that it?

Lincoln-Yuji replied:
We've talked to David Tadokoro. He explained that this means 'substitute patch file for either patchset or series'.

Similar to substitute format using `sed`. Our doubts were clarified

Lincoln-Yuji commented on kw:
```diff
@@ -287,6 +287,13 @@ function kw()
         vm_main "$@"
       )
       ;;
+    handle-trailer)
```
This name is kind of bad in fact. We couldn't come up with a better name while making this feature.

Not sure if this will be put into another already existing feature in KW or if it's going to be a different one. If the latter, then a better name is welcome.

We are open to suggestions.

Lincoln-Yuji commented on src/handle_trailer.sh:
```diff
@@ -0,0 +1,169 @@
+      fi
+    elif is_a_patch "$arg"; then
+      cmd="git interpret-trailers ${arg} --in-place --trailer"
+      cmd+=" \"${TRAILER_TAG} ${TRAILER_VALUE}\""
```
I'm afraid this can't be done, at least we don't know how to build this command correctly without `\"`. The problem with using single-quotes is that we have multiple nested strings. Replacing the `\"` for `'` messes up the cmd's building process and generates commands that are not correctly executed.

We've tried to think of ways to build the command without it, but it messes up the format.

Lincoln-Yuji commented on src/handle_trailer.sh:
```diff
@@ -0,0 +1,169 @@
+    # Check if given argument is either a patch or valid commit reference,
+    # then build the correct command.
+    if [[ $(git cat-file -t "$arg" 2> /dev/null) == 'commit' ]]; then
+      cmd="git commit --quiet --amend --no-edit --trailer \"${TRAILER_TAG} ${TRAILER_VALUE}\""
```
I'm afraid this can't be done, at least we don't know how to build this command correctly without `\"`. The problem with using single-quotes is that we have multiple nested strings. Replacing the `\"` for `'` messes up the cmd's building process and generates commands that are not correctly executed.

We've tried to think of ways to build the command without it, but it messes up the format.

Lincoln-Yuji commented on src/handle_trailer.sh:
```diff
@@ -0,0 +1,169 @@
+      cmd="git commit --quiet --amend --no-edit --trailer \"${TRAILER_TAG} ${TRAILER_VALUE}\""
+      # Only call 'git rebase' if user is trying to handle multiple commits
+      if [[ "$(git rev-parse "$arg")" != "$(git rev-parse HEAD)" ]]; then
+        cmd="git rebase ${arg} --exec '${cmd}' 2> /dev/null"
```
Using `--quiet` does not seem to suppress the following message from `git rebase --exec`:

```
Executing: ...
Executing: git commit --quiet --amend --no-edit --trailer "Name <mail>"
Executing: ...
```

For each commit, the rebase executes the above command and prints that message. This can be visualized when executing the tests. However, redirecting with `2> /dev/null` suppresses this print. Using `git rebase --quiet --exec <cmd>` doesn't seem to solve the issue. There might be a better to solve this though?

Lincoln-Yuji commented on documentation/man/features/kw-handle-trailer.rst:
```diff
@@ -0,0 +1,86 @@
+
+  $ kw handle-trailer -r "Reviewer Name <reviewer@mail.org>" example.patch
+
+  $ kw handle-trailer -a "Some Name <example@mail.com>" example.patch
```
That's a really cool default behavior for this flags. We recently made a force pull with this request in mind and improved the tests to also take this case in consideration.

Lincoln-Yuji commented:
Hello @rodrigosiqueira

We've implemented your suggestions and now we will be waiting for more feedbacks. I noticed you mentioned something about adding `Signed-off-by` as well. Do you think adding a new option such as `--add-signed-off-by | -s` is worth it?

Also, there were somethings related to code style that we didn't manage to change like the `\"` usage and the `2> /dev/null` redirection because it seems that changing those is harder than we expected or maybe we are missing something. They are explained above in this thread.

Lincoln-Yuji commented:
I've asked some suggestions for better names from some friends and ideas they told me were:

- `kw annotate`
- `kw amend`
- `kw tag`
- `kw signature`

Does any of these names sound better or make more sense? I'm also studying the possibility to inject these options into an already implemented feature, but not sure which one and how.

Lincoln-Yuji commented:
Hello @rodrigosiqueira and @davidbtadokoro. I just want to tell you that I will be marking this PR as Draft for now. There are some minor issues I've noted with the code and other additional changes. Also, I'll be renaming this feature to either `annotate` or `signature`, not sure which one yet. I'll try to transition this PR back to Ready to review within the next 2 weeks.

I have intentions to add new options as well:
- `--add-signed-off-by | -s`
- `--add-reported-by | -R`
- `--add-suggested-by | -S`

These long options sound, you know... too long? Don't know if the `add` prefix or the `by` suffix are needed. I'm keeping them
because it sort of acts as documentation. Plus, more experienced users can always use the short options anyway.

Finally, neither Luiza nor Sabrina will keep working on this PR. I will be working on it alone from now, so it might have longer delays to deliver future requested changes. 

If there's any new additional reviews for the current code, you can write them here and I'll work on them as soon as possible.

Lincoln-Yuji commented:
I'm re-opening this PR. Fortunately I had some spare time today to finish making the changes I had in mind.

I believe I managed to fix the issues I found. Also, I added the new option `--add-signed-off-by | -s` I mentioned previously, but I didn't implement any of the other two since they seem to be less necessary to make this feature minimally useful and I'm still not sure if this code will require heavy further changes/refactoring.

I also changed the name to `signature`. Hope this name makes more sense. Waiting for new feedbacks

rodrigosiqueira commented on documentation/man/features/kw-signature.rst:
```diff
@@ -0,0 +1,94 @@
+| *kw* *signature* (-s(<string>) | \--add-signed-off-by=(<string>)) [\--verbose] [<sha> | <patch>]
+| *kw* *signature* (-r(<string>) | \--add-reviewed-by=(<string>)) [\--verbose] [<sha> | <patch>]
+| *kw* *signature* (-a(<string>) | \--add-acked-by=(<string>)) [\--verbose] [<sha> | <patch>]
+| *kw* *signature* (-f(<sha>) | \--add-fixes=(<sha>)) [\--verbose] [<sha> | <patch>]
```
Can we also have:
* Tested-by
* Reported-by
* Co-developed-by

I think with those extra tags, we covered everything :)

rodrigosiqueira commented on documentation/man/features/kw-signature.rst:
```diff
@@ -0,0 +1,94 @@
+===================================================
+kw-signature - Write commit and patch trailer lines
```
I was thinking about the name, and at the end of the day, you are just adding kernel tags to the commit. We can try to find a name around it. Some brainstorm ideas:
* add-commit-tags
* add-tag
* add-signature-tag
* Link -> This is a special case requiring a link, not an email.

rodrigosiqueira replied:
Ok, it is time for us to end this PR. In this sense, let's rename this feature to `kernel-tags`.

rodrigosiqueira commented on documentation/man/features/kw-signature.rst:
```diff
@@ -0,0 +1,94 @@
+'git rebase' and the 'git interpret-trailers' commands. This kw command
+is able to perform usual operations over trailer lines of commits and patches.
+By default, it uses the commit currently pointed by ``HEAD`` as the target
+of the operation if user does not specify one. Also, no operation is used by
+default. At least one operation option must be given.
```
Just add: At least one operation ...

rodrigosiqueira commented on documentation/man/features/kw-signature.rst:
```diff
@@ -0,0 +1,94 @@
+This command does not allow repetition of options. For instance, the
+following command will give an error::
+
+  $ kw signature -r"Reviewer Name <mail>" -r"Some Name <mail>"
```
What if I want to add Reviewed-by from different developers?
For example, I got a Reviewed-by from Joe and another one from Bob, but for some reason, I have to do it manually; why not allow:

```bash
kw signature -r "Joe Doe <somethingxpto@lala.com" -r "Bob Lala <abcd@1234s.com"
```

This scenario looks valid to me. I think you can check for name/email duplication and avoid this case:

```bash
kw signature -r "Joe Doe <somethingxpto@lala.com"  -r "Joe Doe <somethingxpto@lala.com"
```

You can just add a warning in the above case.

Lincoln-Yuji replied:
I've updated the code to allow repetition of options. Though I'm not checking name/email duplication yet.
By default the following command:

```sh
kw signature -r "Joe Doe <somethingxpto@lala.com>"  -r "Joe Doe <somethingxpto@lala.com>"
```

It only writes `Reviewed-by: Joe Doe <somethingxpto@lala.com>` once because it seems that both `git amend --trailer` and `git-interpret-trailers` will not write a duplicate line if the last trailer line is equals to the line it's trying to write right now.

Though it's not hard to come up with ways the user can still somehow force a duplicate line. For example:

```sh
kw signature -r -s -r -s
```

The above command will write:

```
Reviewed-by: Name <Mail>
Signed-off-by: Name <Mail>
Reviewed-by: Name <Mail>
Signed-off-by: Name <Mail>
```

I'm still thinking of ways to solve this issue. Maybe I'll have to, in fact, check all names/emails from each option and check if there are repeating options with the same signature.

rodrigosiqueira replied:
> I've updated the code to allow repetition of options. Though I'm not checking name/email duplication yet. By default the following command:
> 
> ```shell
> kw signature -r "Joe Doe <somethingxpto@lala.com>"  -r "Joe Doe <somethingxpto@lala.com>"
> ```

I suppose the second Joe was a duplication, right? I'll assume you meant Bob.

> 

Now that I'm rereading my comment and checking your reply, I think that a better interface would be:

```sh
kw signature -r "Joe Doe <somethingxpto@lala.com>, Bob Lala <abcd@1234s.com"
```

In other words, just use commas to separate different names.

> It only writes `Reviewed-by: Joe Doe <somethingxpto@lala.com>` once because it seems that both `git amend --trailer` and `git-interpret-trailers` will not write a duplicate line if the last trailer line is equals to the line it's trying to write right now.
> 

I guess this is what we want. However, you can check this yourself in the code and send a warning to the user.

> Though it's not hard to come up with ways the user can still somehow force a duplicate line. For example:
> 
> ```shell
> kw signature -r -s -r -s
> ```
> 

I'm sorry. I guess I was not clear in my original comments. In summary, we don't want duplication of tags with the same name and email. If the tags are different, it is okay to have them, but if they are equal, we don't want them.

> The above command will write:
> 
> ```
> Reviewed-by: Name <Mail>
> Signed-off-by: Name <Mail>
> Reviewed-by: Name <Mail>
> Signed-off-by: Name <Mail>
> ```

From your example, we don't want the duplicated Reviewed-by and Signed-off-by, we just want:

```
Reviewed-by: Name <Mail>
Signed-off-by: Name <Mail>
```

rodrigosiqueira commented on documentation/man/features/kw-signature.rst:
```diff
@@ -0,0 +1,94 @@
+this order, to the commit's trailers pointed by ``HEAD`` using the
+user's git config::
+
+  $ kw signature -r -s
```
In this case, where will it get the name and email information? Git config?

rodrigosiqueira commented on documentation/man/features/kw-signature.rst:
```diff
@@ -0,0 +1,94 @@
+
+  $ kw signature -r -s
+
+Changing the order of the options affects the order the lines are written.
```
Btw, google for this kernel page: Documentation/process/submitting-patches.rst. We can enforce the order suggested by the documentation.

rodrigosiqueira replied:
> I'm very confused about how exactly such order should be enforced. The documentation states that the **Signed-off-by** chain should represent chronologically the real route a patch took as it was propagated. Enforcing an order does not sound feasible to me because it would require the feature to predict the route and interactions between submitters and maintainers.
> 

We don't need to be perfect - maybe just close enough. With this idea in mind, we can try to create a heuristic for dealing with the sequence. For example, you can assume that the dev using the `kw signature` at the moment is the last person in the SoB chain.

> However, there're some details I've noticed while reading this document. It says that every `Co-developed-by` tag indicates authorship, which means that this tag must be immediately followed by a `Signed-off-by` of the same co-author. Since this a general rule, is it ok to implement it such that the option `--add-co-developed-by | -C` also adds an immediate `Signed-off-by` tag? Like:
> 
> ```shell
> $ kw signature -C'Developer Name <developer@mail.xyz>'
> # Written tags
> # Co-developed-by: Developer Name <developer@mail.xyz>
> # Signed-off-by: Developer Name <developer@mail.xyz>
> ```
> 
> Or would using `kw signature -C'blabla' -s'blabla'` still be preferable? 

The option added in your first example is the best way to handle the Co-developed-by tag. However, you might use the second example that you shared, and I think it should not be a problem as well. You just need to handle the duplicated Signed-off-by. 

>Also, it seems that the `Reported-by` flag should be followed by a `Closes` or `Link` tag if the bug report is available in the web. Should this somehow be passed via some argument in the command line?
> 

Keep in mind that in kw, we want to enforce the kernel way of doing things as much as possible. In this sense, we want to incentivize the users to follow the rules, which means we want them to use the Link/Close; however, from my experience, it is common not to have such a link. In this case, we need to define a heuristic. As a first suggestion, we can have something like this:

1. Reported-by might be used with a different (and optional) syntax. Something like this: `kw signature --reported-by "Name <email>;[link|close]=LINK"[,Name <email>...]`.
2. If the user does not provide a link: `kw signature --reported-by "Name <email>"` you can prompt a question asking if they have a link and give the option to insert it in case they forget. If they don't have a link, it should be ok. We probably want to have a force option to avoid asking.

> Sorry if I'm too lost with this. I've been thinking of how to enforce an order according to the documentation but I can't come up with a good idea right now.

rodrigosiqueira commented on documentation/man/features/kw-signature.rst:
```diff
@@ -0,0 +1,94 @@
+=======
+-s, \--add-signed-off-by:
+  Adds a **Signed-off-by** trailer line to either commits or patchsets.
+  By default it uses `user.name` and `user.email` from git config to
+  write these lines if no argument is given to this option.
```
Add this information to the command description.

rodrigosiqueira commented on documentation/man/features/kw-signature.rst:
```diff
@@ -0,0 +1,94 @@
+
+Adding a **Acked-by** line to last 3 commits starting from ``HEAD``::
+
+  $ kw signature --acked-by='Some Name <example@mail.com>' HEAD~3
```
Does this work if I use a SHA commit? For example, in my daily life as a kernel developer, I usually want to add my Signed-off-by from the SHA until the HEAD.

Lincoln-Yuji replied:
I believe what you described is the actual behavior of this feature, though this was not properly documented in the **DESCRIPTION**. I've added an explanation regarding this now in the man page. If there's anything else weirdly presented presented in the documentation please let me know.

rodrigosiqueira commented on documentation/man/features/kw-signature.rst:
```diff
@@ -0,0 +1,94 @@
+
+Adding a **Acked-by** line to last 3 commits starting from ``HEAD``::
+
+  $ kw signature --acked-by='Some Name <example@mail.com>' HEAD~3
```
Also, do the tag options support multiple emails? For example:

```bash
 kw signature --add-reviewed-by='Reviewer Name <reviewer@mail.org>, Another Name <anothername@ljkjsdfal.com>'
```

That results in:

```
Reviewed-by: Reviewer Name <reviewer@mail.org>
Reviewed-by: Another Name <anothername@ljkjsdfal.com>
```

Lincoln-Yuji replied:
Since the command now accepts repetition of options, I'm not sure if supporting multiple signatures in the same argument is required. Also, with the current implementation this will not work beacause the program will accuse `Reviewer Name <reviewer@mail.org>, Another Name <anothername@ljkjsdfal.com>` as an invalid signature format.

However, if we define a **separator**, we can split this argument into multiple signatures and then validate each one.

Supporting multiple emails in the same argument is desirable?

rodrigosiqueira replied:
yeah, and this behavior looks similar to what we have when using git-send-email.

rodrigosiqueira commented on documentation/man/features/kw-signature.rst:
```diff
@@ -0,0 +1,94 @@
+Adding **Fixes** line to commit pointed by ``HEAD``, which fixes another
+previous commit ``90a7ba23340d``::
+
+  $ kw signature --add-fixes=90a7ba23340d
```
Do you know if this command checks if the commit is valid?

rodrigosiqueira commented on documentation/man/features/kw-signature.rst:
```diff
@@ -0,0 +1,94 @@
+
+  $ kw signature --add-fixes=90a7ba23340d
+
+All the above options can be used to perform operations over patches as well::
```
Maybe use `.patch` files.

rodrigosiqueira commented on src/_kw:
```diff
@@ -486,6 +487,17 @@ _kw_self-update()
+    '(-r --add-reviewed-by)'{-r-,--add-reviewed-by=-}'[add Reviewed-by trailer line]' \
+    '(-a --acked-by)'{-a-,--add-acked-by=-}'[add Acked-by trailer line]' \
+    '(-f --add-fixes)'{-f-,--add-fixes=-}'[add Fixes trailer line]' \
+    '1: :_files'
```
@davidbtadokoro could you review this part?

rodrigosiqueira commented on src/signature.sh:
```diff
@@ -0,0 +1,348 @@
+# Define trailer tags
+declare -gA TRAILER_TAGS
+
+TRAILER_TAGS['SIGNOFF']='Signed-off-by:'
+TRAILER_TAGS['REVIEW']='Reviewed-by:'
+TRAILER_TAGS['ACKED']='Acked-by:'
+TRAILER_TAGS['FIXES']='Fixes:'
```
For the keys, let's use the full name, e.g., SIGNED_OFF_BY instead of SINGOF.

rodrigosiqueira commented on src/signature.sh:
```diff
@@ -0,0 +1,348 @@
+function signature_main()
+{
+  local patch_or_sha_args
+
```
Remove this extra new line.

rodrigosiqueira commented on src/signature.sh:
```diff
@@ -0,0 +1,348 @@
+    return 22 # EINVAL
+  fi
+
+  # Writting the trailers respecting the defined order
```
Consider moving this code to its own function.

rodrigosiqueira commented on src/signature.sh:
```diff
@@ -0,0 +1,348 @@
+#
+# Return:
+# In case of successful return 0, otherwise, return 22.
+function parse_signature_options()
```
move this function and the help one to the end of this file.

rodrigosiqueira commented on src/signature.sh:
```diff
@@ -0,0 +1,348 @@
+# an error if they are not properly set with git config.
+#
+# @keyword Holds a string that is a keyword for a specific operation.
+# @signature Holds a string like 'NAME <EMAIL>' defining the signature.
```
Don't forget to document the return option for this function.

rodrigosiqueira commented on src/signature.sh:
```diff
@@ -0,0 +1,348 @@
+{
+  local keyword="$1"
+  local signature="$2"
+
```
Remove this extra new line.

rodrigosiqueira commented on src/signature.sh:
```diff
@@ -0,0 +1,348 @@
+#
+# @keyword Holds a string that is a keyword for a specific operation.
+# @signature Holds a string like 'NAME <EMAIL>' defining the signature.
+function register_signature_operation()
```
I don't think this is a good name. Basically, you are formatting the input, so maybe you can use a name that better describes this.

rodrigosiqueira commented on src/signature.sh:
```diff
@@ -0,0 +1,348 @@
+    if [[ "$?" -gt 0 ]]; then
+      return 1
+    fi
+    options_values["$signature_key"]+="$formated_output"
+  else
+    options_values["$signature_key"]+="$(str_strip "$signature")"
```
You can't trust the input; you should have some level of validation, such as checking for the string size. You can also check if the email has some minimum stuff such as one @ and at least one .; I don't think you need to implement this, iirc, we already have an email validation code.

rodrigosiqueira replied:
We can take this route but add a TODO comment to refactor this later. Ideally, we want to extract that function from a generic place like `kw_string.sh`.

rodrigosiqueira commented on src/signature.sh:
```diff
@@ -0,0 +1,348 @@
+# a valid commit reference. If it is, then it outputs the
+# appropriate formatted message. Else it returns an error code.
+#
+# @sha Holds either a commit hash or pointer
```
Again, don't forget to document the return (review all comments in this PR).

rodrigosiqueira commented on src/signature.sh:
```diff
@@ -0,0 +1,348 @@
+  local sha="$1"
+  local formatted_message
+
+  # Check if given value is a valid commit reference
```
Nice :) This answers my previous question.

rodrigosiqueira commented on src/signature.sh:
```diff
@@ -0,0 +1,348 @@
+  local formatted_message
+
+  # Check if given value is a valid commit reference
+  if [[ $(git cat-file -t "$sha" 2> /dev/null) != 'commit' ]]; then
```
btw, it would be awesome if you could work on the below issue since you wrote a lot of git code:

https://github.com/kworkflow/kworkflow/issues/1123

rodrigosiqueira commented on src/signature.sh:
```diff
@@ -0,0 +1,348 @@
+  local user_name
+  local user_email
+
+  user_name="$(git config user.name)"
+  user_email="$(git config user.email)"
```
@matheustavares do you think we have a better way to get this information?

matheustavares replied:
No, I think that's the intended way for scripts too :)

rodrigosiqueira reviewed (changes requested):
Hi @Lincoln-Yuji ,
First of all, sorry for my huge delay in responding to this change. However, I'm glad that you insisted on it because this feature has great potential and is awesome!

I added many comments and suggestions on how we can proceed with this feature. I did not review the unit test because I think there are already many comments.

Btw, would you consider working on this issue https://github.com/kworkflow/kworkflow/issues/1123 ?

Lincoln-Yuji commented:
Thank you @rodrigosiqueira for the review! I'll take some time to address all these changes. I'll try getting it done within 2 weeks.

I'll keep the name `signature` for now since it sounds low priority compared to everything else. But I'll be revisiting this topic in the future of course.

Plus, I'm not sure if I will have the time to work on issue #1123 right now, but I can take a look at it no problem. Since it's a new library, I believe documenting, testing and effectively applying it into multiple places in the project might be very demanding. I can try opening a new PR in the future to address this issue, but I'll be focusing on this PR for now and then give this **git lib** a shot. Sounds like a natural development, unless this PR can not be accepted without such library already implemented.

rodrigosiqueira commented:
> Thank you @rodrigosiqueira for the review! I'll take some time to address all these changes. I'll try getting it done within 2 weeks.
> 
> I'll keep the name `signature` for now since it sounds low priority compared to everything else. But I'll be revisiting this topic in the future of course.
> 
> Plus, I'm not sure if I will have the time to work on issue #1123 right now, but I can take a look at it no problem. Since it's a new library, I believe documenting, testing and effectively applying it into multiple places in the project might be very demanding. I can try opening a new PR in the future to address this issue, but I'll be focusing on this PR for now and then give this **git lib** a shot. Sounds like a natural development, unless this PR can not be accepted without such library already implemented.

Don't worry, it was just a suggestion. If you can finish this feature, it is more than enough.

rodrigosiqueira commented on src/signature.sh:
```diff
@@ -0,0 +1,402 @@
+      cmd="git commit --quiet --amend --no-edit --trailer \"${trailer}\""
+      # Only call 'git rebase' if user is trying to write multiple commits
+      if [[ "$(git rev-parse "$patch_or_sha")" != "$(git rev-parse HEAD)" ]]; then
+        cmd="git rebase ${patch_or_sha} --exec '${cmd}' 2> /dev/null"
```
Instead of using /dev/null, can we just use `--quiet`.

Lincoln-Yuji replied:
I've tried this before, but the `git rebase` command prints multiple lines like:
```
Executing: git commit --quiet --amend --no-edit --trailer "Tag: Joe Doe <mail>"
Executing: git commit --quiet --amend --no-edit --trailer "Tag: Bob Doe <mail>"
Executing: git commit --quiet --amend --no-edit --trailer "Tag: Michael Doe <mail>"
```
Even when we use `--quiet`, but redirecting `2` to `/dev/null` solves it. I don't know why it behaves like this to be honest and I didn't take the time to investigate it properly. I'll be coming back and taking a look at this in the future, since this is something that has been bothering me as well since I wrote this code.

Lincoln-Yuji replied:
This can be tested running the following command inside a git project:
```bash
git rebase HEAD~3 --quiet --exec "printf 'blabla' > /dev/null" 
```
You will see that even with `--quiet`, the command still outputs the following:
```text
Executing: printf 'blabla' > /dev/null
Executing: printf 'blabla' > /dev/null
Executing: printf 'blabla' > /dev/null
```

rodrigosiqueira replied:
Interesting. In that case, we can live with that. Could you add a tiny comment about that to prevent someone from changing it in the future?

rodrigosiqueira commented on src/signature.sh:
```diff
@@ -0,0 +1,402 @@
+        cmd="git rebase ${patch_or_sha} --exec '${cmd}' 2> /dev/null"
+      fi
+    elif is_a_patch "$patch_or_sha"; then
+      cmd="git interpret-trailers ${patch_or_sha} --in-place --trailer \"${trailer}\""
```
I didn't know this command before, but I just realized that this is a powerful tool. I thought we could use it with the `--where` parameter to define some heuristics. For example, I really want that at least one Signed-off-by be at the end when we use tags like Tested-by, Reviewed-by, etc. This week, when I used your feature to add the tag `tested-by`, it was added at the end; if it had been added before the last Signed-off-by, it would have been perfect for my use case.

Lincoln-Yuji replied:
The `--where` option completely went under my radar. Looks like a nice tool to use. Unfortunately this seems to be limited to `git interpret-trailers` while `git commit --amend` can't benefit from it. However, I've been working on some heuristics to sort the tags **before** writing them using this command, but as you mentioned all new trailers are written at the end by default and only `interpret-trailers` can change this behavior. I'll be revisiting this topic later because this is a very powerful feature in fact.

I've looked up some solutions to work with commits but all of them comes down to using `sed` in very obscure ways, directly editing the commit messages automatically with `rebase` and `amend`. But I'm too afraid to try this, because I have no idea if I'll be able to come up with a good way to implement that, and also it sounds extremely dangerous.

rodrigosiqueira commented on src/signature.sh:
```diff
@@ -0,0 +1,402 @@
+
+        if [[ "$return_status" -eq 1 ]]; then
+          options_values['ERROR']='You must configure your user.name and user.email with git '
+          options_values['ERROR']+='to use --add-signed-off-by or -s without an argument'
```
Just use `--add-singed-off-by|-s`.

rodrigosiqueira commented on src/signature.sh:
```diff
@@ -0,0 +1,402 @@
+  # If user doesn't have either a name or email configured with
+  # git then they must provide an argument
+  if [[ -z "$user_name" || -z "$user_email" ]]; then
+    return 1
```
How about return 61 (ENODATA - No data available), instead of 1.

rodrigosiqueira commented on src/signature.sh:
```diff
@@ -0,0 +1,402 @@
+    trailer+="$formated_output"
+  else
+    signature="$(str_strip "$signature")"
+    if [[ "$keyword" != 'FIXES' ]] && ! is_valid_signature "$signature"; then
```
Wait... why is FIXES a factor here but not in the above ones? I did not get it.

Lincoln-Yuji replied:
The FIXES factor is not necessary above because the `--add-fixes` options demands an argument, which means that it's not possible (or at least it shouldn't) for the user to call this option with empty content. When I wrote this code I assumed that if `"$signature"` is not given then the caller is not trying to parse a `Fixes` tag line, unless the developer working on this feature (in that case, me) messed up and forgot to pass the second parameter.

In this line we need the FIXES factor because this tag's content follows a different format and it was already validated and parsed previously in `format_fixes_message`.

I'm open to suggestions on how to improve this code's flow though. I agree it might be confusing.

rodrigosiqueira commented on documentation/man/features/kw-signature.rst:
```diff
@@ -0,0 +1,112 @@
+
+  $ kw signature -a'Some Name <example@mail.org>' example.patch
+
+  $ kw signature -f'90a7ba23340d' example.patch
```
btw, when I tried -f in a specific branch, I got this error:
![Screenshot from 2024-07-27 10-54-58](https://github.com/user-attachments/assets/ad724175-a1fd-4408-9cb4-79871b7666f8)

Lincoln-Yuji replied:
I don't know how to reproduce this error, but if no warning nor errors were reported by the feature itself before reaching this sequence of git errors then I'm almost sure the command assembled in `write_all_trailers` is getting messed up somehow.

Does using `--verbose` prints anything unusual/weird? Or nothing happens at all? By looking at the command used it should print something like:
```bash
git commit --quiet --amend --no-edit --trailer "Fixes: 51fc8079c75e (\"<message>\")"
```

Lincoln-Yuji replied:
Ohh ok, I think I understand that now. Each error is associated to a token in the commit message. The message is being expanded by the shell and causing that to happen when `cmd_manager` evals the command. Can you please give the message that is forcing those errors. I believe I overlooked some important details while making this code...

rodrigosiqueira replied:
I used this repository:
https://gitlab.freedesktop.org/agd5f/linux
The commit SHA is: 51fc8079c75ea89500a2953d02dcd304719f6929

The commit message is:
```
commit 51fc8079c75ea89500a2953d02dcd304719f6929
Author: Asad Kamal <asad.kamal@amd.com>
Date:   Mon Jan 8 18:34:55 2024 +0800

    Revert "drm/amdgpu: Update irq disable flow during unload"
    
    Revert the commit as the original workaround:
    commit f5c7e7797060 ("drm/amdgpu: Adjust removal control
    flow for smu v13_0_2") that caused the issue needs re-evaluation
    
    This reverts commit 2e8e792e6a51e8cb7f5f96148146b6525dbb9cef.
    
    Signed-off-by: Asad Kamal <asad.kamal@amd.com>
    Reviewed-by: Lijo Lazar <lijo.lazar@amd.com>
```

Lincoln-Yuji commented:
Hello @rodrigosiqueira, thank you for your new feedbacks. I have been very busy recently so I'm sorry for the long wait.
I too have noticed that the biggest issue so far is lack of tests for git branch setups or bad inputs, since I also had some of these issues while making further tests. I'll try making this feature more robust to avoid such annoying problems.

Lincoln-Yuji commented on src/signature.sh:
```diff
@@ -0,0 +1,402 @@
+# In case of successful return 0, otherwise, return 22.
+function parse_signature_options()
+{
+  local long_options='add-signed-off-by::,add-reviewed-by::,add-acked-by::,add-fixes::'
+  long_options+=',add-tested-by::,add-reported-by::,add-co-developed-by::,verbose'
+  local short_options='s::,r::,a::,f::,t::,R::,C::'
```
That actually happened to me multiple times when I was testing it myself. I had to use `git reset --hard <SHA>` to backup every time that happened. I've been thinking of how to properly handle this, but didn't have too much free time recently to work on it. But I'm going to try come up with ideas to improve robustness and avoid this really annoying issue.

Lincoln-Yuji replied:
The solution I'll be using for now is giving an error to the user and stop the feature from running any further if an invalid commit or patch is detected within the arguments. Not sure if it's the best solution, but at least it will help us to avoid this catastrophic scenario where many undesirable changes are made and the user might not have a proper way to back it all up.

Lincoln-Yuji commented:
I'll try making a mini changelog for some changes so far:
- Moved email validation functions from `send_patch` to `kwlib` so they can be used by other features.
- Support for more kernel tags: `Tested-by`, `Co-developed-by` and `Reported-by`.
- New tags are sorted before getting added to the commit or patch (some details already included in the man page). However, they are all added at the end of old tags already present. Maybe we can use `--where` for **patchsets** specifically since `interpret-trailers` offers this very nice tool.
- If an argument is neither a valid commit nor a patch, then the feature stops running and prompts an error, avoiding the annoying and dangerous scenario when we might forget to use `=` delimiter for long options or to put the argument glued to the short options (e.g `--add-fixes='123abc123'` and `-f'123abc123'`).

Lincoln-Yuji commented:
Sorry for the additional force-pushes. Some unit tests were failing in the CI but locally they were all passing. So I did this to test if something on my branches were possibly out of date even after synchronizing my repo branches with upstream.

rodrigosiqueira commented on documentation/man/features/kw-signature.rst:
```diff
@@ -0,0 +1,172 @@
+
+.. code-block:: text
+
+   $ kw signature -s'Joe Doe <joe@mail.xyz>' -s -t -t'Joe Doe <joe@mail.xyz>'
```
Maybe move -s and -t to the beginning, it is weird to see it in the middle. Or this was intentional?

rodrigosiqueira commented on kw:
```diff
@@ -286,6 +286,13 @@ function kw()
         vm_main "$@"
       )
       ;;
+    signature)
```
As I commented in the documentation part, let's rename this feature now.

rodrigosiqueira reviewed (changes requested):
Hi @Lincoln-Yuji,

Sorry for the delay again. I just added a few more comments, but aside from that, let's start to wrap up this feature; we can fix other issues later.

rodrigosiqueira commented:
Hi @Lincoln-Yuji , why did you close this PR? Is it because you don't want to work on it anymore?
